src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1598,8 +1598,12 @@ bool match_global_settings(struct fi_info * prov)
      * but as more and more providers supported this feature, the decision was made to
      * require it. */
     CHECK_CAP(MPIDI_OFI_ENABLE_TAGGED,
-              !(prov->caps & FI_TAGGED) || !(prov->caps & FI_DIRECTED_RECV) ||
-              prov->domain_attr->cq_data_size < 4);
+              !(prov->caps & FI_TAGGED) || prov->domain_attr->cq_data_size < 4);
+
+    /* OFI provider doesn't expose FI_DIRECTED_RECV by default for performance consideration.
+     * MPICH should request this flag to enable it. */
+    if (MPIDI_OFI_ENABLE_TAGGED)
+        prov->caps |= FI_DIRECTED_RECV;
```

Comment:
What does it mean to add a capability here without checking it against the provider? I don't quite understand what this change is doing.

Reply:
This is a change to align with libfaric 1.9.1 release. From 1.9.1, libfabric will hide FI_DIRECTED_RECV flag in default capabilities flags even providers supports it. The reason is enabling FI_DIRECTED_REV has performance impact.

MPICH use fi_getinfo(0, NULL) to query available providers on system firstly, and if checking FI_DIRECTED_RECV against that, providers hiding FI_DIRECTED_REV (even they actually support it) will be filtered out. 

Based on the feedback from libfabric team, all providers support FI_TAG will support FI_DIRECTED_RECV also. So this change moves FI_DIRECTED_RECV into application requested flag when MPICH wants OFI tag support.

@aingerson please corrects above comments if I'm wrong.

Reply:
> Based on the feedback from libfabric team, all providers support FI_TAG will support FI_DIRECTED_RECV also. So this change moves FI_DIRECTED_RECV into application requested flag when MPICH wants OFI tag support.

Is this true of releases prior to v1.9.1?

Reply:
We have a testing system using "verbs;ofi_rxm" provider, which already has this issue.

Reply:
This commit comes from the fix of running MPICH on OFI for that "verbs;ofi_rxm" system.

Reply:
Libfabric version is 1.9.0 on that system.

Reply:
My question is, do all providers supporting FI_TAGGED also support FI_DIRECTED_RECV in prior versions of libfabric? This change does not differentiate based on library version. It is a potential compatibility break.

Reply:
@aingerson Can you confirm this question?

We verified sockets, psm2, and verbs;ofi_rxm on our side.

Reply:
The only RDM provider I could see being a problem is psm. We're not testing/maintaining psm so I can't speak on its status.
This shouldn't make it not work with providers that don't support both. It would just make sure that if MPICH is trying to use tagged, that they correctly request FI_DIRECTED_RECV as well because they require it when using the tagged path. We can always fall back to the normal msg path if the provider doesn't support it. 
A full fix for providers that did not support both before would be to try to request FI_DIRECTED_RECV and then if that also returned ENODATA then to fallback to not using tagged. As is, the workaround for providers that don't support both in earlier versions would be to force tagged off using the environment variable.

Reply:
If libfabric is going to such lengths to hide this capability because of its poor performance, it seems like a mistake for MPICH to put it in the default capability set, doesn't it? What do we lose if we remove it? Tag and rank space?

Reply:
FI_DIRECTED_RECV refers to rank match in MPICH.

Reply:
The default capability set is TAG, we firstly check TAG requirement for FI_TAG and cq_data_sz>=4 to get candidate providers. Then we add FI_DIRECTED_RECV into hints to turn on rank match for chosen provider. It still works, right?

Reply:
OK, I looked at the rxm provider code and I think better understand. A refresh of our 2-step provider selection code also helped. I'm OK with this going in.